plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,284 @@
+                if content is None:
+                    shutdown = True
+                    continue
+
+                for declarative_artifact in content.d_artifacts:
+                    one_artifact_q = Q()
+                    for digest_name in declarative_artifact.artifact.DIGEST_FIELDS:
+                        digest_value = getattr(declarative_artifact.artifact, digest_name)
+                        if digest_value:
+                            key = {digest_name: digest_value}
+                            one_artifact_q &= Q(**key)
+                    if one_artifact_q:
+                        all_artifacts_q |= one_artifact_q
+
+            for artifact in Artifact.objects.filter(all_artifacts_q):
+                for content in batch:
+                    if content is None:
+                        continue
+                    for declarative_artifact in content.d_artifacts:
+                        for digest_name in artifact.DIGEST_FIELDS:
+                            digest_value = getattr(declarative_artifact.artifact, digest_name)
+                            if digest_value and digest_value == getattr(artifact, digest_name):
+                                declarative_artifact.artifact = artifact
+            for content in batch:
+                if content is None:
+                    continue
+                await out_q.put(content)
+            batch = []
+            if shutdown:
+                break
+        await out_q.put(None)
+
+
+class ArtifactDownloader(BaseStage):
+    """
+    A Stages API stage to download :class:`~pulpcore.plugin.models.Artifact` files, but don't save
+    the :class:`~pulpcore.plugin.models.Artifact` in the db.
+
+    This stage downloads the file for any :class:`~pulpcore.plugin.models.Artifact` objects missing
+    files and creates a new :class:`~pulpcore.plugin.models.Artifact` object from the downloaded
+    file and its digest data. The new :class:`~pulpcore.plugin.models.Artifact` is not saved but
+    added to the :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object, replacing the likely
+    incomplete :class:`~pulpcore.plugin.models.Artifact`.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage creates a ProgressBar named 'Downloading Artifacts' that counts the number of
+    downloads completed. Since it's a stream the total count isn't known until it's finished.
+
+    This stage drains all available items from `in_q` and starts as many downloaders as possible.
+
+    Args:
+        max_concurrent_downloads (int): The maximum number of concurrent downloads this stage will
+            run. Default is 100.
+    """
+
+    def __init__(self, max_concurrent_downloads=100, *args, **kwargs):
+        self.max_concurrent_downloads = max_concurrent_downloads
+        super().__init__(*args, **kwargs)
```

Comment:
It's better practice to initialize objects in heirarchical order.  Please call `super().__init__()` (first) before setting extended attributes.

Reply:
ok will do in next push